### PR TITLE
[PLAT-7635] check nil parameter in network breadcrumbs

### PR DESCRIPTION
**Pull request by kstenerud**
Apple metrics via the Apple API might not have a request object, or the request might not have a URL. We need to check before trying to use them.

Also made the request search more robust since there can be many possible places for them, and not all of them will have an HTTP body.

## Testing

Unfortunately we don't have a repro case for the report that triggered this issue, but the existing tests all pass.

### CODE REVIEW

File: BugsnagNetworkRequestPlugin/BugsnagNetworkRequestPlugin/BSGURLSessionTracingDelegate.m [hunk 1]
```diff
@@ -83,12 +83,32 @@ + (nonnull NSString *)URLStringWithoutQueryForComponents:(nonnull NSURLComponent
...
+                return req;
+            }
+        }
+        // Failing that, just use the last phase.
+        req = transactionMetrics[transactionMetrics.count - 1].request;
+    }
+
+    return req;
+}
+
```
Comment (nickdowell) [inline BugsnagNetworkRequestPlugin/BugsnagNetworkRequestPlugin/BSGURLSessionTracingDelegate.m:108]: Why does this logic need to be so complicated?

If we have no way to reproduce a nil `originalRequest`, is it even worth having untested logic?

Comment (kstenerud) [reply BugsnagNetworkRequestPlugin/BugsnagNetworkRequestPlugin/BSGURLSessionTracingDelegate.m:108]: I wanted to try harder to get a valid request. The `transactionMetrics` array can contain a whole bunch of different objects depending on what phase of the request it was in (including retries from a 3xx response). If you think it's too risky I can just go back to the old way.

Comment (nickdowell) [reply BugsnagNetworkRequestPlugin/BugsnagNetworkRequestPlugin/BSGURLSessionTracingDelegate.m:108]: The NSURLSession header says:

```
@interface NSURLSessionTask : NSObject <NSCopying, NSProgressReporting>

@property (readonly)                 NSUInteger    taskIdentifier;    /* an identifier for this task, assigned by and unique to the owning session */
@property (nullable, readonly, copy) NSURLRequest  *originalRequest;  /* may be nil if this is a stream task */
@property (nullable, readonly, copy) NSURLRequest  *currentRequest;   /* may differ from originalRequest due to http server redirection */
@property (nullable, readonly, copy) NSURLResponse *response;         /* may be nil if no response has been received */
```

So we may be getting a nil `originalRequest` if it was a "stream task" - does that give a lead to find a way to reproduce the crash?

If so and this logic can't find a request for that case, probably better to keep it simple like:

```
        NSURLRequest *req = task.originalRequest ?: task.currentRequest;
        if (!req) {
            return;
        }
```

Comment (kstenerud) [reply BugsnagNetworkRequestPlugin/BugsnagNetworkRequestPlugin/BSGURLSessionTracingDelegate.m]: OK, I've put the simplified version in. I couldn't repro the crash using a stream task unfortunately.

File: BugsnagNetworkRequestPlugin/BugsnagNetworkRequestPlugin.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -757,7 +757,7 @@
 			buildSettings = {
 				ASSETCATALOG_COMPILER_APPICON_NAME = AppIcon;
 				CODE_SIGN_STYLE = Automatic;
-				DEVELOPMENT_TEAM = 372ZUL2ZB7;
+				DEVELOPMENT_TEAM = "";
```
Comment (nickdowell) [inline BugsnagNetworkRequestPlugin/BugsnagNetworkRequestPlugin.xcodeproj/project.pbxproj:760]: These changes should not be committed

File: BugsnagNetworkRequestPlugin/BugsnagNetworkRequestPlugin/BSGURLSessionTracingDelegate.m [hunk 3]
```diff
@@ -97,9 +94,20 @@ - (void)URLSession:(__unused NSURLSession *)session task:(NSURLSessionTask *)tas
...
         metadata[@"duration"] = @((unsigned)(metrics.taskInterval.duration * 1000));
-        metadata[@"method"] = req.HTTPMethod;
-        metadata[@"url"] = [BSGURLSessionTracingDelegate URLStringWithoutQueryForComponents:urlComponents];
-        metadata[@"urlParams"] = [BSGURLSessionTracingDelegate urlParamsForQueryItems:urlComponents.queryItems];
+
+        NSURLRequest *req = task.originalRequest;
+        if (req == nil) {
+            req = task.currentRequest;
+        }
+        if (req) {
```
Comment (nickdowell) [inline BugsnagNetworkRequestPlugin/BugsnagNetworkRequestPlugin/BSGURLSessionTracingDelegate.m:102]: If there's no request (and hence no url) is there any point leaving a breadcrumb at all?

@kattrali

Comment (kstenerud) [reply BugsnagNetworkRequestPlugin/BugsnagNetworkRequestPlugin/BSGURLSessionTracingDelegate.m:102]: Better some info than none?

**Review by nickdowell — APPROVED** [review]

**Pull request merged**